### ISSUE TITLE
Implement retry logic for Kinesis APIs

### ISSUE THREAD

**Issue by guilload** [quickwit-oss/quickwit#751]
Retries have been excluded from the current implementation because the retry logic used for S3 requires a little refactor to be reused elsewhere.

I left a bunch of `TODO` where the retry logic needs to be plugged in.

**Comment by guilload** [comment]
@dai-dao, are you actively working on this? Feel free to ping me on Discord for additional guidance.

**Comment by boraarslan** [comment]
There are a few ways to implement retry logic here:

1. The retry logic on the S3 side could be refactored to be used on the Kinesis side
2. A new retry logic can be implemented.

Current retry logic is not really customizable and would be insufficient if different policies are needed for different scenarios. In that case, a new logic can be implemented. There are customizable retry policies on other AWS SDKs ([Java](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/retry/RetryPolicy.RetryPolicyBuilder.html), [JS](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-kinesis/classes/adaptiveretrystrategy.html), [Python](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html)) but as far as I know, `rusoto` doesn't have request retry functionality. 

[Boto3's documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html) explains the retry policy used is consistent with other AWS SDKs, so a new implementation might be written based on that. The new retry implementation can be used on the S3 side too but writing it from the scratch might be unnecessary if such functionality would not be needed in the future.

After your feedback, I can work on this issue.

**Comment by fmassot** [comment]
Thanks @boraarslan. Can you be more accurate on the retry policy parameters? What do you propose exactly in solution 1 to make the retry policy work on S3 and Kinesis? Knowing that we would want different parameters? Ideally, you can write some pseudo code to have a good idea of what you have in mind.

**Comment by boraarslan** [comment]
The current `retry` function needs the given Future's Error type to implement `E: IsRetryable + Display + 'static`. On the S3 side `RusotoError`s are wrapped in `RusotoErrorWrapper<T>` to satisfy the requirements. To reuse the already implemented version both retry logic and Error wrappers can be moved elsewhere (I am not very familiar with the codebase so I don't know exactly where).and then on the Kinesis side requests can be run exactly like on the S3 side:

```rust
let response = retry(|| async {
        kinesis_client
            .get_records(request.clone())
            .await
            .map_err(RusotoErrorWrapper::from)
    })
    .await?;
```

One thing to note is that function return type `anyhow::Result` is used on the Kinesis side whereas on the S3 side, `StorageResult` is used. Looking at the todo comments, reusing the `RusotoErrorWrapper` and writing `From` impls for future custom `KinesisError`(?) type seems reasonable to reuse existing code.

The problem with the `retry` function is like I said in my previous comment, it doesn't support different retry policies(like [Legacy][1], [Standard][2], [Adaptive][3]) and it is not customizable for different clients or API calls. It just supports exponential backoff for retriable errors.

A possible solution might be creating a trait(RetryRequestTrait(?)) that would hide the retry policy and creating a struct that implements the trait to execute the request and hide Error wrapping logic. Struct is initialized with the request function, executes the request, and casts the errors behind the scenes.

```mermaid
    flowchart TD;
    A[RetryRequestTrait] -- implemented by --> B[S3 Executor Struct];
    A -- implemented by --> C[Kinesis Executor Struct];
    B -- returns --> D[StorageResult];
    C -- returns --> E[`anyhow::Error` or future custom error like KinesisError];
```
As for pseudocode, RetryRequestTrait might look something like this:

```rust
trait RetryRequestTrait {
    pub async fn request() -> Result; // No default impl. execute_request runs this function
    pub async fn execute_request(RetryPolicy) -> Result {
        // If request is retriable
        match RetryPolicy {
            RetryPolicy::Legacy => execute_legacy(),
            RetryPolicy::Standard => execute_standard(),
            RetryPolicy::Adaptive => execute_adaptive(),
        }
    }
    pub async fn execute_legacy() -> Result {/* default impl here */}
    pub async fn execute_standard() -> Result {/* default impl here */}}
    pub async fn execute_adaptive() -> Result {/* default impl here */}}
}
```

and structs might look something like this:

```rust
struct KinesisRequestExecutor(request);

impl RetryRequestTrait for KinesisRequestExecutor {
    pub async fn request() -> Result {
        /* execute request */
    }
}

impl KinesisRequestExecutor {
    /* Calling trait functions and Error casting/wrapping logic here */
}
```

This way retry policies could be passed to the struct per request to override the default retry policy.
This solution might be longer to implement but considering Kinesis API would need its custom Error type in the future, this also makes it possible to use already written Error logic.

(I am not confident with the names, naming could be better. Also, I am not very familiar with the project layout so I can't propose where to move the existing code.)

This solution was the first one that came to my mind which makes sense but I might be mistaken on some of the topics. I can rethink another solution or modify this one based on your feedback.

[1]: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.htmllegacy-retry-mode
[2]: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html#standard-retry-mode
[3]: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html#adaptive-retry-mode

**Comment by boraarslan** [comment]
I created a PR (#1451) that implements a very basic `RetryRequest` struct. It is a slightly modified version of the retry login on the S3 side with some modifiable parameters. It currently doesn't have any retry policy other than exponential backoff. I also initialized the struct on some of the requests to check compile errors and it seems ok for now. I can implement other policies too if you want to.

**Comment by guilload** [comment]
> The current `retry` function needs the given Future's Error type to implement `E: IsRetryable + Display + 'static`. On the S3 side `RusotoError`s are wrapped in `RusotoErrorWrapper<T>` to satisfy the requirements. To reuse the already implemented version both retry logic and Error wrappers can be moved elsewhere (I am not very familiar with the codebase so I don't know exactly where).and then on the Kinesis side requests can be run exactly like on the S3 side:
> 
> ```rust
> let response = retry(|| async {
>         kinesis_client
>             .get_records(request.clone())
>             .await
>             .map_err(RusotoErrorWrapper::from)
>     })
>     .await?;
> ```
We like this approach, which keeps things simple and maximizes code reuse. Do we have to move the common code to a new workspace package for things to work or, since Kinesis lives in the indexing package, which depends on storage, can we get away with making some functions and structs public in the storage package? 

> One thing to note is that function return type `anyhow::Result` is used on the Kinesis side whereas on the S3 side, `StorageResult` is used. Looking at the todo comments, reusing the `RusotoErrorWrapper` and writing `From` impls for future custom `KinesisError`(?) type seems reasonable to reuse existing code.

Yes, the retry function should return a `RusotoErrorWrapper`, and the caller should convert that error to whatever is required.
 
> The problem with the `retry` function is like I said in my previous comment, it doesn't support different retry policies(like [Legacy](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.htmllegacy-retry-mode), [Standard](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html#standard-retry-mode), [Adaptive](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/retries.html#adaptive-retry-mode)) and it is not customizable for different clients or API calls. It just supports exponential backoff for retriable errors.

There might have been a communication issue here. We're perfectly happy with one retry strategy for now (exponential backoff with jitter). What we want is the ability to customize the parameters of the strategy. What that in mind, can you design a simple solution to do so without introducing a new trait?

**Comment by guilload** [comment]
Closed by #1451
